common/params.cc

```diff
@@ -326,3 +328,29 @@ void Params::clearAll(ParamKeyType key_type) {
 
   fsync_dir(getParamPath());
 }
+
+struct AsyncWriter {
+  void queue(const std::tuple<std::string, std::string, std::string> &dat) {
+    q.push(dat);
+    // start thread on demand
+    if (!f.valid() || f.wait_for(std::chrono::milliseconds(0)) == std::future_status::ready) {
+      f = std::async(std::launch::async, &AsyncWriter::write, this);
+    }
+  }
+
+  void write() {
+    std::tuple<std::string, std::string, std::string> dat;
+    while (q.try_pop(dat, 0)) {
+      auto &[path, key, value] = dat;
+      Params(path).put(key, value);
+    }
+  }
+
+  std::future<void> f;
+  SafeQueue<std::tuple<std::string, std::string, std::string>> q;
+};
+
+void Params::putNonBlocking(const std::string &key, const std::string &val) {
+  static AsyncWriter async_writer;
```

Comment:
why static instead of part of the class?

Reply:
There may be many instances of Params. Only one asycwriter is needed per process.